### PR TITLE
Additional OpenSSL compat stuff for OpenSSH

### PR DESCRIPTION
**Pull request by julek-wolfssl**
- group->curve_nid is now set to the real NID of the curve
- Fix memory leak with EncryptDerKey

New functions implemented
- `EC_POINT_is_on_curve`
- `i2d_EC_PUBKEY`
- `i2d_ECPrivateKey`
- `wc_ecc_point_is_on_curve`
- `wolfSSL_i2d_PUBKEY`
- `wolfSSL_i2d_PrivateKey`
- `wolfSSL_EC_GROUP_dup`
- `wolfSSL_d2i_ECPrivateKey`
- `wolfSSL_EC_POINT_add`
- `wolfSSL_EC_POINT_invert`
- `wolfSSL_d2i_X509_NAME`

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -32113,8 +32134,11 @@ int wolfSSL_EC_KEY_generate_key(WOLFSSL_EC_KEY *key)
         return 0;
     }
 
+    /* NIDToEccEnum returns -1 for invalid NID so if key->group->curve_nid
+     * is 0 then pass 0 as arg */
     if (wc_ecc_make_key_ex(rng, 0, (ecc_key*)key->internal,
-                                        key->group->curve_nid) != MP_OKAY) {
+            key->group->curve_nid ? NIDToEccEnum(key->group->curve_nid) : 0
```
Comment (SparkiDev) [inline src/ssl.c:32140]: Move the ternary out of the function call - use a new variable.

File: src/ssl.c [hunk 2]
```diff
@@ -32910,11 +32942,103 @@ int wolfSSL_i2o_ECPublicKey(const WOLFSSL_EC_KEY *in, unsigned char **out)
...
+    }
+
+    return len;
+}
+
 void wolfSSL_EC_KEY_set_conv_form(WOLFSSL_EC_KEY *eckey, char form)
 {
-    if (eckey && (form == POINT_CONVERSION_COMPRESSED ||
-                  form == POINT_CONVERSION_UNCOMPRESSED)) {
+    if (eckey && (form == POINT_CONVERSION_COMPRESSED
```
Comment (SparkiDev) [inline src/ssl.c:33034]: POINT_CONVERSION_UNCOMPRESSED?

File: src/ssl.c [hunk 3]
```diff
@@ -33089,6 +33236,70 @@ int wolfSSL_EC_POINT_set_affine_coordinates_GFp(const WOLFSSL_EC_GROUP *group,
...
+        WOLFSSL_MSG("mp_read_radix prime error");
+        goto cleanup;
+    }
+
+    if (mp_montgomery_setup(&prime, &mp) != MP_OKAY) {
+        WOLFSSL_MSG("mp_montgomery_setup nqm error");
+        goto cleanup;
+    }
+
+    if (ecc_projective_add_point((ecc_point*)p1->internal, (ecc_point*)p2->internal,
```
Comment (SparkiDev) [inline src/ssl.c:33284]: The points being added must be in Montgomery form.
Is this always true?

File: src/ssl.c [hunk 4]
```diff
@@ -37057,6 +37305,42 @@ static int CopyX509NameToCertName(WOLFSSL_X509_NAME* n, CertName* cName)
...
+        DecodedCert cert;
+
+        WOLFSSL_ENTER("wolfSSL_d2i_X509_NAME");
+
+        if (!in || !*in || length <= 0) {
+            WOLFSSL_MSG("Bad argument");
+            return NULL;
+        }
+
+        InitDecodedCert(&cert, *in, length, NULL);
```
Comment (SparkiDev) [inline src/ssl.c:37321]: Put a comment in here that you are placing an encoded X509 name as the data but only calling GetName to parse data.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:33034]: Just a typo. Good catch. Thanks!

File: src/ssl.c (hunk 3 above)
Comment (julek-wolfssl) [reply src/ssl.c:33284]: I'm not sure. I based the `wolfSSL_EC_POINT_add` implementation on other uses of `ecc_projective_add_point`. Is there some way I could check if they are in Montgomery form?

Comment (SparkiDev) [reply src/ssl.c:33284]: The point won't have its ordinates in Montgomery form.
You'll need to convert to and from in this function.

To convert to Montgomery, take a look at: ecc.c:2635-2684 (wc_ecc_mulmod_ex())
Map the point back call: ecc_map_ex()

Add a test and maybe even try compiling a simplified test against OpenSSL and wolfSSL and see if you get the same results.

Comment (julek-wolfssl) [reply src/ssl.c:33284]: Does this mean that something similar should be done in `wolfSSL_EC_POINT_mul`?

Comment (SparkiDev) [reply src/ssl.c:33284]: Yes, well, sort of.
ecc_mul2add does everything internally.
The wc_ecc_mulmod shouldn't be mapping the results back to affine and out of Montgomery form. Last parameter is 1 and should be 0 for both calls. The add is expecting Montgomery form. Then the mapping done after (ecc_map) converts the Montgomery form in projective to affine in normal form.
The others cases look right.

**Review by SparkiDev — APPROVED** [review]

File: src/ssl.c [hunk 5]
```diff
@@ -33236,6 +33491,43 @@ int wolfSSL_EC_POINT_mul(const WOLFSSL_EC_GROUP *group, WOLFSSL_EC_POINT *r,
...
+    p = (ecc_point*)a->internal;
+
+
+    /* read the curve prime and a */
+    if (mp_init_multi(&prime, NULL, NULL, NULL, NULL, NULL) != MP_OKAY) {
+        WOLFSSL_MSG("mp_init_multi error");
+        return WOLFSSL_FAILURE;
+    }
+
+    if (mp_sub(&prime, p->y, p->y) != MP_OKAY) {
```
Comment (toddouska) [inline src/ssl.c:33518]: Couldn't p be NULL here?

Comment (julek-wolfssl) [reply src/ssl.c:33518]: Added a NULL check on function entry.

File: src/ssl.c [hunk 6]
```diff
@@ -7614,6 +7615,39 @@ WOLFSSL_EVP_PKEY* wolfSSL_d2i_PUBKEY(WOLFSSL_EVP_PKEY** out,
     return pkey;
 }
 
+/* helper function to get raw pointer to DER buffer from WOLFSSL_EVP_PKEY */
+static int wolfSSL_EVP_PKEY_get_der(const WOLFSSL_EVP_PKEY* key, unsigned char** der)
+{
+    unsigned char* pt;
+    int sz = key->pkey_sz;
```
Comment (toddouska) [inline src/ssl.c:7622]: key may be NULL here.

**Review by toddouska — APPROVED** [review]

**Pull request merged**